UnitsNet.Serialization.JsonNet/UnitsNetJsonConverterBase.cs

```diff
@@ -0,0 +1,130 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+using System;
+using System.Linq.Expressions;
+using Newtonsoft.Json;
+
+namespace UnitsNet.Serialization.JsonNet
+{
+    /// <summary>
+    /// Base class for converting units to/from JSON using Json.Net
+    /// </summary>
+    /// <typeparam name="T">The type of unit to convert. For example, <see cref="UnitsNet.Mass"/>.</typeparam>
+    public class UnitsNetJsonConverterBase<T> : JsonConverter 
+        where T : struct, IComparable, IComparable<T>
+    {
+        private Func<T, double> _valueWriter;
+        private Enum _baseUnit;
+        private Func<double, T> _valueReader;
+
+        /// <summary>
+        /// Determines whether this instance can convert the specified object type.
+        /// </summary>
+        /// <param name="objectType">Type of the object.</param>
+        /// <returns></returns>
+        public override bool CanConvert(Type objectType)
+        {
+            return objectType == typeof(T);
+        }
+
+        /// <summary>
+        /// Initializes the expressions used to store the value to write to JSON and the base unit.
+        /// </summary>
+        /// <param name="expressionToStore">The expression to store the unit value to JSON.</param>
+        /// <param name="expressionToRestore">The expression to restore the unit value from JSON.</param>
+        protected void Init(Expression<Func<T, double>> expressionToStore, Enum baseUnit, Expression<Func<double, T>> restoreExpression)
+        {
+            _valueWriter = expressionToStore.Compile();
+            _baseUnit = baseUnit;
+            _valueReader = restoreExpression.Compile();
+        }
+
+        /// <summary>
+        /// Reads the JSON representation of the object.
+        /// </summary>
+        /// <param name="reader">The <see cref="T:Newtonsoft.Json.JsonReader" /> to read from.</param>
+        /// <param name="objectType">Type of the object.</param>
+        /// <param name="existingValue">The existing value of object being read.</param>
+        /// <param name="serializer">The calling serializer.</param>
+        /// <returns>
+        /// The object value.
+        /// </returns>
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            object value = null;
+            string enumStr = null;
+            while (reader.Read())
+            {
+                if (value != null && enumStr != null)
+                    break;
+
+                switch (reader.TokenType)
```

Comment:
This deserializer seems fragile to me. If I understand correctly, it assumes any JSON float/int is the value and any JSON string is the unit. I believe it would be best to check that the JSON property names are "Value" and "Unit" too?

How is the JSON deserializer supposed to work for Data Transfer Objects like this?
```
{
"Name": "Andreas",
"Height": {
  "Value": 1.84,
  "Unit": "LengthUnit.Meter"
  }
}
```

Do we need to tag the `Height` C# property with a LengthAttribute or similar?